providers/implementations/keymgmt/ecx_kmgmt.c

```diff
@@ -594,8 +607,25 @@ static void *ecx_gen(struct ecx_gen_ctx *gctx)
         ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
         goto err;
     }
-    if (RAND_priv_bytes_ex(gctx->libctx, privkey, key->keylen, 0) <= 0)
-        goto err;
+#ifndef FIPS_MODULE
+    if (gctx->dhkem_ikm != NULL) {
+        if (gctx->type == ECX_KEY_TYPE_ED25519
+                || gctx->type == ECX_KEY_TYPE_ED448)
+            goto err;
+        if (!ossl_ecx_dhkem_derive_private(key, privkey,
+                                           gctx->dhkem_ikm, gctx->dhkem_ikmlen))
+            goto err;
+        if (!ossl_ecx_public_from_private(key))
+            goto err;
+        key->haspubkey = 1;
```

Comment:
@mattcaswell moving the code into here shows that it is not doing the masking on the private key buffer as is done further down in this code block.. Is this a problem?

Reply:
What is the masking below good for? Is it required for the X*** private keys to work properly?

Reply:
See https://www.rfc-editor.org/rfc/rfc7748 section 5:

>    Scalars are assumed to be randomly generated bytes.  For X25519, in
   order to decode 32 random bytes as an integer scalar, set the three
   least significant bits of the first byte and the most significant bit
   of the last to zero, set the second most significant bit of the last
   byte to 1 and, finally, decode as little-endian.  This means that the
   resulting integer is of the form 2^254 plus eight times a value
   between 0 and 2^251 - 1 (inclusive).  Likewise, for X448, set the two
   least significant bits of the first byte to 0, and the most
   significant bit of the last byte to 1.  This means that the resulting
   integer is of the form 2^447 plus four times a value between 0 and
   2^445 - 1 (inclusive).

Reply:
So, is the output from `ossl_ecx_dhkem_derive_private` just a random 32 byte string? In which case it will need to be masked. Or has it already been masked?

Reply:
The description listed at https://www.rfc-editor.org/rfc/rfc9180.html#name-derivekeypair does not do this..
The test vectors listed in the RFC (e.g https://www.rfc-editor.org/rfc/rfc9180.html#name-base-setup-information) do not have the masking..
e.g.

ikmE: (The seed)
7268600d403fce431561aef583ee1613527cff655c1343f29812e66706df3234
pkEm: (public bytes)
37fda3567bdbd628e88668c3c8d7e97d1d1253b6d4ea6d44c150f741f1bf4431
skEm: (private bytes)
52c4a758a802cd8b936eceea314432798d5baf2d7e9235dc084ab1b9cfa2f736

Reply:
There is another section [7.1.2. ](https://www.rfc-editor.org/rfc/rfc9180.html#section-7.1.2)[SerializePrivateKey and DeserializePrivateKey](https://www.rfc-editor.org/rfc/rfc9180.html#name-serializeprivatekey-and-des) that does this by the looks of it..

Reply:
Not sure those test vectors are correct.. As it says they are in serialized form,,

Reply:
Probably the masking needs to be done for these keys too.

Reply:
Added. The tests need to do the masking in order to counter a deficiency in the test vectors.